### DISCUSSION TITLE
Ninja-Multiconfig

### DISCUSSION THREAD

**Discussion by Zingam** [microsoft/vscode-cmake-tools#2764] [Q&A] [answered]
I'm trying to figure out how Ninja-Multiconfig works. Is it possible to select a Debug, Release etc. config to build and run?

I always get this:

      [cpptools] The build configurations generated do not contain the active build configuration. Using "Debug" for CMAKE_BUILD_TYPE instead of "null" to ensure that IntelliSense configurations can be found

**Comment by bobbrow** [comment, marked as answer]
I believe this commit fixes it: https://github.com/microsoft/vscode-cmake-tools/commit/ebb08ce02762249dea40a3c4bffa87e105682f2e
It should be in 1.13.10

**Reply by Zingam** [reply]
It looks like that I need to create a CMake Build preset:

```CMake
    "buildPresets": [
        {
            "name": "default.ninja-multi.Debug",
            "description": "Ninja-Multiconfig Debug",
            "displayName": "Debug",
            "configurePreset": "default.ninja-multi",
            "configuration": "Debug"
        },
        {
          "name": "default.ninja-multi.Release",
          "description": "Ninja-Multiconfig Release",
          "displayName": "Release",
          "configurePreset": "default.ninja-multi",
          "configuration": "Release"
      }
    ]
```

It is not clear to me if it is possible to have such  presets created automatically by default? Or maybe it's enough to document it.

I switched to the pre-release version 1.13.12. Also clicking on `Build` causes these messages to popup:

![image](https://user-images.githubusercontent.com/3337419/194706429-b2033785-b73b-4d19-9f63-ceeb7f3025a7.png)

**Reply by Zingam** [reply]
If we have custom build presets, do we need VS to add a default one too by default?

**Reply by bobbrow** [reply]
When you write `configurePresets` that use a multi-config generator (like Ninja Multi-Config or Visual Studio), you have to create `buildPresets` for the `CMAKE_BUILD_TYPE`s that you want to be able to build because all of them will be generated in the same "build" folder when you configure and we don't know which one to pick.

VS's CMake Project template uses the single config Ninja generator so it does not need `buildPresets`

Those errors appear to be some problem with the tree view in the Project Outline?  If you have a repro you can share, please open an issue for it.

**Reply by Zingam** [reply]
> VS's CMake Project template uses the single config Ninja generator so it does not need `buildPresets`

A noob question. I decided to try Ninja-Multiconfig. It turned out the default build preset doesn't support it properly.

Would it make sense to document that a Ninja multiconfig needs build presets and give an example.

Also maybe the default build preset option could provide an extended build presets with Debug, Release in case the generator is a Ninja multiconfig or at least issue a warning? Would that make sense too?

**Reply by Zingam** [reply]
I must admit that when I have noticed that cmake tools support Ninja multiconfig I decided to check what it is without knowing anything about it.

**Reply by bobbrow** [reply]
> Would it make sense to document that a Ninja multiconfig needs build presets and give an example.
> 
> Also maybe the default build preset option could provide an extended build presets with Debug, Release in case the generator is a Ninja multiconfig or at least issue a warning? Would that make sense too?

These are all good ideas.  If you would like to open issues for them, we'll work on them when we get some time, or accept PR's from the community.
